### ISSUE TITLE
issues creating a google_compute_security_policy with rules created in dynamic blocks

### ISSUE THREAD

**Issue by abby-embree-airship** [hashicorp/terraform-provider-google#14178]
<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

<!--- Thank you for keeping this note for the community --->

### Terraform Version

```
Terraform v1.4.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/google v4.59.0
+ provider registry.terraform.io/hashicorp/google-beta v4.59.0
```

### Affected Resource(s)

* google_compute_security_policy


### Terraform Configuration Files

creating a security policy with rules generated via dynamic block does not work as expected; in some cases (i think relating to importing an existing rule then trying to change it vs creating a new one via a terraform apply) i get an error, other times the terraform plan simply shows no changes even when the list the dynamic block is iterating over via for_each changes - create the resource, add an entry to the list used to create rule blocks dynamically, then run a new plan, and you'll see no changes even though you should get a plan that includes adding the new rule block. example tf code:

```tf
      resource "google_compute_security_policy" "this" {                                                                                     
        name  = "test"                                                                                                          
                                                                                                                                                  
        dynamic "rule" {                                                                                                                          
          for_each = local.block                                                                                                    
          iterator = block                                                                                                                        
                                                                                                                                                  
          content {                                                                                                                               
            action   = "deny(502)"                                                                                                                
            priority = sum([block.key, 2])                                                                                                        
            match {                                                                                                                               
              versioned_expr = "SRC_IPS_V1"                                                                                                       
              config {                                                                                                                            
                src_ip_ranges = block.value                                                                                             
              }                                                                                                                                   
            }                                                                                                                                     
            description = "test"                                                      
          }                                                                                                                                       
        }                                                                                                                                         
                                                                                                                                                  
        rule {                                                                                                                                    
          action   = "allow"                                                                                                                      
          priority = "2147483647"                                                                                                                 
          match {                                                                                                                                 
            versioned_expr = "SRC_IPS_V1"                                                                                                         
            config {                                                                                                                              
              src_ip_ranges = ["*"]                                                                                                               
            }                                                                                                                                     
          }                                                                                                                                       
          description = "default allow"                                                                                                           
        }                                                                                                                                         
      }
```


### Error Output

the errors i encountered trying to change the number of rules on an imported resource:

```
╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.temporary-cloudarmor-rules.google_compute_security_policy.temporary to include new values learned so far
│ during apply, provider "registry.terraform.io/hashicorp/google" produced an invalid new value for .rule: planned set element
│ cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("allow"), "description":cty.StringVal("Default rule, higher priority overrides it"),
│ "header_action":cty.ListValEmpty(cty.Object(map[string]cty.Type{"request_headers_to_adds":cty.List(cty.Object(map[string]cty.Type{"header_name":cty.String,
│ "header_value":cty.String}))})),
│ "match":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"config":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"src_ip_ranges":cty.SetVal([]cty.Value{cty.StringVal("*")})})}),
│ "expr":cty.ListValEmpty(cty.Object(map[string]cty.Type{"expression":cty.String})), "versioned_expr":cty.StringVal("SRC_IPS_V1")})}),
│ "preview":cty.False, "priority":cty.NumberIntVal(2.147483647e+09),
│ "rate_limit_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"ban_duration_sec":cty.Number,
│ "ban_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number})), "conform_action":cty.String,
│ "enforce_on_key":cty.String, "enforce_on_key_name":cty.String, "exceed_action":cty.String,
│ "exceed_redirect_options":cty.List(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String})),
│ "rate_limit_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number}))})),
│ "redirect_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String}))}) does not correlate with any
│ element in actual.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.temporary-cloudarmor-rules.google_compute_security_policy.temporary to include new values learned so far
│ during apply, provider "registry.terraform.io/hashicorp/google" produced an invalid new value for .rule: planned set element
│ cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("deny(403)"), "description":cty.StringVal(""),
│ "header_action":cty.ListValEmpty(cty.Object(map[string]cty.Type{"request_headers_to_adds":cty.List(cty.Object(map[string]cty.Type{"header_name":cty.String,
│ "header_value":cty.String}))})),
│ "match":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"config":cty.ListValEmpty(cty.Object(map[string]cty.Type{"src_ip_ranges":cty.Set(cty.String)})),
│ "expr":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"expression":cty.StringVal("has(request.headers['user-agent']) &&
│ request.headers['user-agent'].contains('Nuclei') \n")})}), "versioned_expr":cty.StringVal("")})}), "preview":cty.False,
│ "priority":cty.NumberIntVal(1), "rate_limit_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"ban_duration_sec":cty.Number,
│ "ban_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number})), "conform_action":cty.String,
│ "enforce_on_key":cty.String, "enforce_on_key_name":cty.String, "exceed_action":cty.String,
│ "exceed_redirect_options":cty.List(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String})),
│ "rate_limit_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number}))})),
│ "redirect_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String}))}) does not correlate with any
│ element in actual.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.temporary-cloudarmor-rules.google_compute_security_policy.temporary to include new values learned so far
│ during apply, provider "registry.terraform.io/hashicorp/google" produced an invalid new value for .rule: planned set element
│ cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("deny(403)"), "description":cty.StringVal(""),
│ "header_action":cty.ListValEmpty(cty.Object(map[string]cty.Type{"request_headers_to_adds":cty.List(cty.Object(map[string]cty.Type{"header_name":cty.String,
│ "header_value":cty.String}))})),
│ "match":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"config":cty.ListValEmpty(cty.Object(map[string]cty.Type{"src_ip_ranges":cty.Set(cty.String)})),
│ "expr":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"expression":cty.StringVal("has(request.headers['user-agent']) &&
│ request.headers['user-agent'].contains('nuclei')")})}), "versioned_expr":cty.StringVal("")})}), "preview":cty.False,
│ "priority":cty.NumberIntVal(2), "rate_limit_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"ban_duration_sec":cty.Number,
│ "ban_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number})), "conform_action":cty.String,
│ "enforce_on_key":cty.String, "enforce_on_key_name":cty.String, "exceed_action":cty.String,
│ "exceed_redirect_options":cty.List(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String})),
│ "rate_limit_threshold":cty.List(cty.Object(map[string]cty.Type{"count":cty.Number, "interval_sec":cty.Number}))})),
│ "redirect_options":cty.ListValEmpty(cty.Object(map[string]cty.Type{"target":cty.String, "type":cty.String}))}) does not correlate with any
│ element in actual.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```

### Expected Behavior

changing the content of the list being for_each'ed in the dynamic block should result in planned changes to the resource


### Actual Behavior

either nothing happens (erroneously empty plan), or an error like the one above occurs

**Comment by edwardmedia** [comment]
@abby-embree-airship dynamic is a Terraform feature and I can't really commit much on that. But before you apply any content changes, did you see any diffs in the plan after the import? If yes, you probably need to fix that first before moving to the next step. Maybe to start without dynamic?

**Comment by edwardmedia** [comment]
@abby-embree-airship before you adding the dynamic code, do you see an issue with its plan?

**Comment by abby-embree-airship** [comment]
no, but that doesn't mean that the error is indicative of an issue with dynamic blocks rather than the provider - see https://github.com/hashicorp/terraform/issues/28340

**Comment by edwardmedia** [comment]
@abby-embree-airship using below config, I tested apply -> import -> apply on the imported config. No issue is found.  What should I do in order for me to repro the issue?

```tf
resource "google_compute_security_policy" "policy" {

  name = "my-policy"

  rule {
    action   = "deny(403)"
    priority = "1000"
    match {
      versioned_expr = "SRC_IPS_V1"
      config {
        src_ip_ranges = ["9.9.9.0/24"]
      }
    }
    description = "Deny access to IPs in 9.9.9.0/24"
  }

  rule {
    action   = "allow"
    priority = "2147483647"
    match {
      versioned_expr = "SRC_IPS_V1"
      config {
        src_ip_ranges = ["*"]
      }
    }
    description = "default rule"
  }
}
```

**Comment by abby-embree-airship** [comment]
the config you posted won't reproduce the issue because it doesn't create rules via dynamic blocks. the issue i linked indicates that problems with creating dynamic blocks within certain resource types may be indicative of a provider-level issue rather than an hcl-level issue because of how different resource types handle dynamic blocks. I'm working on a complete example to demonstrate this issue, will post when finished

**Comment by edwardmedia** [comment]
@rileykarson what do you think about this issue?

**Comment by abby-embree-airship** [comment]
here's some example code that demonstrates this issue - a less complicated conditional in constructing `valid_blocks` won't reproduce the problem:

```
locals {
  blocklist = [
    {
      ip          = "192.0.0.8"
      description = "test 1"
	  block_until = "1970-01-01T00:00:00Z"
      # block_until = "2024-01-01T00:00:00Z"
    },
    {
      ip          = "8.8.8.8"
      description = "test 2"
      # block_until = "1970-01-01T00:00:00Z"
      block_until = "2024-01-01T00:00:00Z"
    }
  ]
  valid_blocks = [
    for block in local.blocklist : {
      ip          = block.ip
      description = block.description
    }
    if timecmp(timestamp(), block.block_until) == -1
  ]
}

resource "google_compute_security_policy" "this" {
  name  = "test"

  dynamic "rule" {
    for_each = local.valid_blocks
    iterator = block

    content {
      action   = "deny(502)"
      priority = block.key
      match {
        versioned_expr = "SRC_IPS_V1"
        config {
          src_ip_ranges = [block.value.ip]
        }
      }
      description = block.value.description
    }
  }

  rule {
    action   = "allow"
    priority = "2147483647"
    match {
      versioned_expr = "SRC_IPS_V1"
      config {
        src_ip_ranges = ["*"]
      }
    }
    description = "default allow"
  }
}
```

do the following:

* apply the above terraform code
* check the cloudarmor rule created in GCP, you should see the rule with 1 entry with a description like "test 2"
*  swap the dates by commenting lines 6 and 13 and uncommenting lines 7 and 12
* run a terraform plan - this will show no planned changes even though the contents of `local.valid_blocks` has just changed
* destroy the resource
* terraform apply again
* check the cloudarmor rule in GCP again, description of the rule should now be "test 1"

I think this has something to do with the unusually vague plan output seen below:

```
  # google_compute_security_policy.this will be created
  + resource "google_compute_security_policy" "this" {
      + fingerprint = (known after apply)
      + id          = (known after apply)
      + name        = "test"
      + project     = (known after apply)
      + self_link   = (known after apply)
      + type        = (known after apply)
    }
```

again, I think the root of this vagueness has to do with the problem described in https://github.com/hashicorp/terraform/issues/28340 . also, I haven't tested a function call other than timecmp in the for loop creating `local.valid_blocks`, but i know that having a boolean compare instead of comparing the output of a function doesn't reproduce the issue. the above code below for example works fine:

```
  blocklist = [
	{
	  ip           = "192.0.0.8"
	  should_block = true
	},
	{
	  ip           = "8.8.8.8"
	  should_block = false
	}
  ]
  valid_blocks = [
	for block in local.blocklist : block.ip
	if block.should_block
  ]
```

**Comment by abby-embree-airship** [comment]
as for the provider error i posted, i don't remember the exact circumstances that produced that error, but i will work on making it happen again. it's related to this issue somehow, but is tangential to the main problem addressed above

**Comment by abby-embree-airship** [comment]
@edwardmedia update your terraform version, the timecmp function only exists in terraform 1.3 and later. I used 1.4.0 when writing up this issue. see https://developer.hashicorp.com/terraform/language/functions/timecmp

**Comment by edwardmedia** [comment]
@abby-embree-airship I realized the version. After updating it to 1.4.5, it works now.

**Comment by abby-embree-airship** [comment]
try making the change i mentioned - switch the block_until dates, then re-run the plan. the expected results would be that there would be changes in the plan, the fact that there aren't is a bug.

**Comment by edwardmedia** [comment]
@abby-embree-airship I see what you meant now. But if we ignore `valid_blocks` and use the `blocklist` directly, any changes in the locals like `description` trigger the a diff in the plan which seems correct. To me, this is still a question or issue in the terraform core. I do not see a problem in the provider. @SarahFrench  what do you think?  

```tf
locals {
  blocklist = [
    {
      ip          = "192.0.0.8"
      description = "test 1"
      # block_until = "1970-01-01T00:00:00Z"
      block_until = "2024-01-01T00:00:00Z"
    },
    {
      ip          = "8.8.8.8"
      description = "test 2"
      block_until = "1970-01-01T00:00:00Z"
      # block_until = "2024-01-01T00:00:00Z"
    }
  ]
  valid_blocks = [
    for block in local.blocklist : {
      ip          = block.ip
      description = block.description
    }
    if timecmp(timestamp(), block.block_until) == -1
  ]
}

resource "google_compute_security_policy" "this" {
  name  = "issue14178-user"

  dynamic "rule" {
    //for_each = local.valid_blocks
    for_each = local.blocklist. <------ loop blocklist directly
    iterator = block

    content {
      action   = "deny(502)"
      priority = block.key
      match {
        versioned_expr = "SRC_IPS_V1"
        config {
          src_ip_ranges = [block.value.ip]
        }
      }
      description = block.value.description
    }
  }

  rule {
    action   = "allow"
    priority = "2147483647"
    match {
      versioned_expr = "SRC_IPS_V1"
      config {
        src_ip_ranges = ["*"]
      }
    }
    description = "default allow"
  }
}
```

**Comment by abby-embree-airship** [comment]
i believe this to be a provider issue because of what's discussed in https://github.com/hashicorp/terraform/issues/28340, but also because we can test whether this behavior is replicable across different resource types. consider this resource:

```
resource "null_resource" "this" {
  triggers = {
    test = jsonencode(local.valid_blocks)
  }
}
```

when i create it with the blocklist above, then switch around the commented lines and run a plan, i get this:

```
  # null_resource.this must be replaced
-/+ resource "null_resource" "this" {
      ~ id       = "8952275810646135394" -> (known after apply)
      ~ triggers = { # forces replacement
          ~ "test" = jsonencode(
                [
                  - {
                      - description = "test 2"
                      - ip          = "8.8.8.8"
                    },
                ]
            ) -> (known after apply)
        }
    }
```

this tells me the problem isn't with the `timecmp` function or with the code that's creating `valid_blocks`. I'm still working on finding a resource to use as a test where i can create dynamic blocks that will accept content from `valid_blocks`, will post an update once i have found something, but i expect it will behave correctly for other resource types and not for this specific resource type.

**Comment by abby-embree-airship** [comment]
okay, here's an example of this exact approach working as expected with a different resource type. hopefully this should demonstrate that this issue is specific to this provider and resource type rather than an issue with core terraform:

```
resource "google_sql_database_instance" "db" {
  name             = "test-deleteme"
  database_version = "POSTGRES_11"
  region           = "us-east1"

  settings {
    tier = "db-f1-micro"

    ip_configuration {
      ipv4_enabled = true
      require_ssl  = true

      dynamic "authorized_networks" {
        for_each = local.valid_blocks
        iterator = block

        content {
          name  = block.value.description
          value = "${block.value.ip}/32"
        }
      }
    }
  }
}
```

apply this, then switch the commented out dates, then plan - you will see proposed changes. apply it and you'll see that the authorized networks for the db changes from `192.0.0.8` to `8.8.8.8`. this replicates the same dynamic block functionality as i used for the security policy, but with a different resource type

**Comment by abby-embree-airship** [comment]
also, i just tested this with a condition other than testing the output of `timecmp` - `if length(regexall("/24/", block.block_until)) > 0` replicates this same behavior

**Comment by SarahFrench** [comment]
Hi @abby-embree-airship 👋  - thanks for your reproduction steps in your [previous comment](https://github.com/hashicorp/terraform-provider-google/issues/14178#issuecomment-1501367734)! It helped me familiar with this issue quickly.

## Problem

I think I've found the underlying cause of the issue - the `timestamp()` built-in function in Terraform core. This function is not called during plan steps and is only called during apply steps. For more info please see this, quite hard to spot, [note in the documentation](https://developer.hashicorp.com/terraform/language/expressions/function-calls#when-terraform-calls-functions).

You can explore this by altering the locals block in your reproduction instructions to use this new `local.current_time` value and un/comment different values.

```hcl
locals {
  blocklist = [
    {
      ip          = "192.0.0.8"
      description = "test 1"
      block_until = "2024-01-01T00:00:00Z"
    },
    {
      ip          = "8.8.8.8"
      description = "test 2"
      block_until = "1970-01-01T00:00:00Z"
    }
  ]
  current_time = timestamp() # Use when initially applying the config
  # current_time = "2023-04-18T00:00:00Z" # 1 in valid_blocks - Uncomment when playing with plans after initial apply
  # current_time = "2050-01-01T00:00:00Z" # 0 in valid_blocks - Uncomment when playing with plans after initial apply
  # current_time = "1950-01-01T00:00:00Z" # 2 in valid_blocks - Uncomment when playing with plans after initial apply
  valid_blocks = [
    for block in local.blocklist : {
      ip          = block.ip
      description = block.description
    }
    if timecmp(local.current_time, block.block_until) == -1
  ]
}
```

In plan steps (following the initial apply to create the `google_compute_security_policy` resource) I believe Terraform is unable to detect that a change is needed because it doesn't have access to timestamp()'s value, therefore doesn't know what local.valid_blocks's value is, so doesn't know how to make a diff between old and new values of local.valid_blocks.

---

## Ideas for next steps

I can see your use case is depending on Terraform to generate a plan to update the `google_compute_security_policy` resource whenever a block_until time swaps from being in the future to being in the past. Currently the timestamp() issue blocks this completely I'm afraid 😬

First, I think it'd be useful if you [opened an issue in the core Terraform repo](https://github.com/hashicorp/terraform/issues/new/choose) as I think your use case is totally valid.

In terms of helping you achieve a policy that updates correctly as time passes, one crude option if you are using Terraform OSS is to run CLI commands with a `-replace` flag ([see here in docs](https://developer.hashicorp.com/terraform/cli/commands/plan#replace-address)). This is not the ideal solution, but if you force Terraform to replace the resource it'll reassess the contents of `local.valid_blocks` and recreate the policy in a way that respects the current time.

Another option comes to mind that _would work_ but has a lot of potential to be mis-used; you could replace use of timestamp() with use of a provider that allowed you to access the value of [time.Now()](https://pkg.go.dev/time#Now) in RFC3339 format as a data source. Data sources are refreshed at plan time, so it returns the correct current time. I think the best solution though is for Terraform core to support 'plan-time timestamps' or another solution that is less easy to misuse.

---

Phew, sorry for writing so much! Please let me know if you have any further questions - I'm happy to discuss this more if needed, but I think we'll eventually close this issue as the problem lies with TF core and not the Google provider

**Comment by abby-embree-airship** [comment]
so i was looking through the code a little and it seems like the root of the issue may be [this check](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_compute_security_policy.go#L658), which compares the old and new key values provided by https://pkg.go.dev/github.com/hashicorp/terraform/helper/schema#ResourceData.GetChange - a call that happens a few lines above at https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_compute_security_policy.go#L631 . the issue with doing this comparison is that i believe that `oSet` and `nSet` _might not contain anything at all_, or could be in a state that does not reflect the actual state of the proposed delta, due to something mentioned in [the link you posted previously](https://developer.hashicorp.com/terraform/language/expressions/function-calls#when-terraform-calls-functions), which indicates that the use of these functions produces a 'known after apply' state AKA an unknown value. this is mentioned in other parts of the documentation as being problematic; `customizeDiff`, for example, doesn't support known after apply values: 'NOTE: CustomizeDiff does not currently support computed/"known after apply" values from other resource attributes.' (taken from https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/customizing-differences). i found this issue https://github.com/hashicorp/terraform-plugin-sdk/issues/1040 discussing how an unknown value coming from something like our dynamic block might behave that has the line 'The map value doesn't include the "Unknown" element at all' in it, and i'm definitely not an expert in the internal workings of terraform and don't actually know go at all but i think based on all this information that something like this might be happening in the code (taken from `resourceComputeSecurityPolicyUpdate` https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_compute_security_policy.go#L572 comments added by me):

```
func resourceComputeSecurityPolicyUpdate(d *schema.ResourceData, meta interface{}) error {
	config := meta.(*Config)

        //...

	if d.HasChange("rule") {
		o, n := d.GetChange("rule") // getchange returns a map containing no rules due to SDK defect 1040
		oSet := o.(*schema.Set)
		nSet := n.(*schema.Set)

		oPriorities := map[int64]bool{}
		nPriorities := map[int64]bool{}
		for _, rule := range oSet.List() {
			oPriorities[int64(rule.(map[string]interface{})["priority"].(int))] = true
		}

               // i think if you logged oPriorities and nPriorities here you would maybe see something interesting, including maybe two empty maps

		for _, rule := range nSet.List() { // no iterations of this for loop run
			priority := int64(rule.(map[string]interface{})["priority"].(int))
			nPriorities[priority] = true
			if !oPriorities[priority] {
				client := config.NewComputeClient(userAgent)

				// If the rule is in new and its priority does not exist in old, then add it.
				op, err := client.SecurityPolicies.AddRule(project, sp, expandSecurityPolicyRule(rule)).Do() // this is why this behavior doesn't result in existing rules in the remote object being deleted, because rule changes are additive instead of building a list of desired rules, wiping out all rules in the remote object, then pushing a new set of rules determined in the previous step

				if err != nil {
					return errwrap.Wrapf(fmt.Sprintf("Error updating SecurityPolicy %q: {{err}}", sp), err)
				}

				err = ComputeOperationWaitTime(config, op, project, fmt.Sprintf("Updating SecurityPolicy %q", sp), userAgent, d.Timeout(schema.TimeoutUpdate))
				if err != nil {
					return err
				}
			} else if !oSet.Contains(rule) {
				// ...
			}
		}

		for _, rule := range oSet.List() { // doesn't trigger if oSet is empty
                        // ...
		}
	}

	return resourceComputeSecurityPolicyRead(d, meta)
}
```

i'm not super confident about all this, but a quick way to check whether i'm on to something or not would be to log the contents of `oSet` and `nSet` and see if there is anything strange going on there. like i said i don't know go or any provider development stuff so i can't really test it myself. if it turns out that i'm correct about the source of this issue, then the solution would be to migrate to V2 of the plugin SDK mentioned (here)[https://developer.hashicorp.com/terraform/plugin/framework/migrating/benefits]. that page's subheading named ''null-unknown-and-known-values" [doesn't actually exist anywhere in the doc](https://developer.hashicorp.com/terraform/plugin/framework/migrating/benefits#null-unknown-and-known-values), i assume they forgot to add it, but i'm guessing that the fact that it was supposed to be included means that the V2 SDK at least includes hints in the schema data model that indicate when there are null or unknown values so that they can be handled properly or at least more gracefully than the V1 model where those values are just silently absent.

@edwardmedia @SarahFrench

**Comment by SarahFrench** [comment]
Hey, thanks for looking into this. I was doing some similar hunting through the code for a smoking gun yesterday and I found something that might interest you.

First I want to acknowledge some of the points you raised above:

>so i was looking through the code a little and it seems like the root of the issue may be [this check](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_compute_security_policy.go#L658)

Unfortunately the Update/Create/Delete functions are only invoked after a plan has been made identifying that those actions need to be taken on a resource. During the plan stage only Read is invoked, when the state is refreshed. The things relevant  to whether a plan is made or not are Terraform core, the SDK, the schema of the resource, and the Read function.

>solution would be to migrate to V2 of the plugin SDK mentioned [here](https://developer.hashicorp.com/terraform/plugin/framework/migrating/benefits).

There is currently a project starting to migrate this provider to the plugin framework and there are lot of benefits I'm looking forward to! The SDK is known to have a lot of quirks that the DevEx team has learned from when making this framework.

---

Now, thinking about the different behaviours of dynamic blocks in `google_compute_security_policy` versus `google_sql_database_instance`. I found that the reason why we can generate plans for the DB instance while using cynamic blocks  is that [the `authorized_networks` field does not contain `Computed: true`](https://github.com/hashicorp/terraform-provider-google/blob/ac0ee3012045d768da917840a76b9c141b361600/google/resource_sql_database_instance.go#L370-L375). The `rules` block in the [security policy resource does have  `Computed: true` in the schema](https://github.com/hashicorp/terraform-provider-google/blob/ac0ee3012045d768da917840a76b9c141b361600/google/resource_compute_security_policy.go#L67) and this prevents the a diff being identified and a plan being generated. I found that if I built the Google provider locally with the resource_compute_security_policy.go#L67 line removed then I could get the `google_compute_security_policy` resource to behave like `google_sql_database_instance` when generating plans for dynamic blocks. Similarly, if I made the `authorized_networks` field `Computed: true` I can make it no longer able to make plans for that block.

The reason for this is a quirk of the SDK - it can conflate Computed and Unknown values. I agree that a 'more correct' behaviour for the `google_compute_security_policy` resource would for the plan to be generated each time. However we can't make the rule block not Computed as that would be a breaking change for other users. There doesn't seem to be a correct way to make the resource that caters for everyone; the SDK's problems can block certain things being done.

So the ways to solve this issue seem to be:
- Somehow make the current timestamp a known value, instead of unknown
    - [plantimestamp](https://github.com/hashicorp/terraform/pull/32980) will help here. The docs in that PR say it'll be released in Terraform 1.5
- Resolve the SDK issues by migrating to the plugin framework.
   - Migration work is currently underway - the provider currently combines the framework and the old SDK
   -  I'm afraid I don't have a timeline I can share with you for when _this_ resource will be migrated

I appreciate this is annoying, but I don't think there are any bug fixes that can be made quickly to address this issue.

---

One last thing, a possible way to "make the current timestamp a known value" without waiting on a HashiCorp release is to use a provider that returns the current time as a data source. Data sources are refreshed during plan generation (assuming there are no dependencies with other resources) and are known values. I have toyed with making one on my laptop and I could share it with you if you wish?

**Comment by SarahFrench** [comment]
Ah the bash script is much simpler than a whole provider, nice! I'm glad we could get your use case working 😄
